pyroclast/prototype/model.py

```diff
@@ -0,0 +1,117 @@
+                 prototype_dim,
+                 num_classes,
+                 class_specific=0):
+        """
+        Args:
+            conv_stack (tf.Module): Convolutional network which ends in a 7x7xC representation where C is arbitrary, but typically equal to the prototype_dim
+            num_prototypes (int): Number of prototype vectors in the model
+            prototype_dim (int): Dimensionality of each prototype vector
+            num_classes (int): Number of classes in the classification problem
+            class_specific (bool): Optional, whether to use class-specific prototypes and objective
+        """
+        self.prototype_dim = prototype_dim
+        self.class_specific = class_specific
+
+        self.conv_stack = conv_stack
+        self.final_conv = tf.keras.Sequential([
+            tf.keras.layers.Conv2D(prototype_dim,
+                                   1,
+                                   1,
+                                   'same',
+                                   activation=tf.nn.relu),
+            tf.keras.layers.Conv2D(prototype_dim,
+                                   1,
+                                   1,
+                                   'same',
+                                   activation=tf.nn.sigmoid)
+        ])
+        self.prototype_layer = PrototypeLayer(num_prototypes, prototype_dim)
+        self.max_pool = tf.keras.layers.GlobalMaxPool2D()
+
+        if class_specific:
+            # set prototype class identity
+            num_prototypes_per_class = num_prototypes // num_classes
+            self.prototype_class_identity = np.zeros(
+                [num_prototypes, num_classes], dtype=np.float32)
+            for j in range(num_prototypes):
+                self.prototype_class_identity[j, j //
+                                              num_prototypes_per_class] = 1
+
+            # create classifier
+            classifier_kernel_init = lambda shape, dtype: -0.5 * tf.ones(
+                shape, dtype=dtype) + 1.5 * self.prototype_class_identity
+            self.classifier = tf.keras.layers.Dense(
+                num_classes,
+                use_bias=False,
+                kernel_initializer=classifier_kernel_init)
+        else:
+            self.classifier = tf.keras.layers.Dense(num_classes, use_bias=False)
+
+    def __call__(self, x):
+        """
+        Args:
+            x (Tensor): image data to be classified
+
+        Returns:
+            Tuple of tensors (y_hat, minimum distances)
+        """
+        conv_output = self.final_conv(self.conv_stack(x))
+        assert conv_output.shape[1:3] == [7, 7]
+        distances, similarities = self.prototype_layer(conv_output)
+        minimum_distances = -self.max_pool(-distances)
+        prototype_activations = self.max_pool(similarities)
+        return self.classifier(prototype_activations), minimum_distances
+
+    def conv_prototype_objective(self, min_distances, label=None):
+        """
```

Comment:
What exactly are you trying to optimize here? You can add equations to the docstring, so it would be nice to know what is being optimized here (without referring to the paper).

Reply:
Also, might be a good place to include the multi-line comment you have later on.